src/SystemState.ts

```diff
@@ -101,4 +106,45 @@ export class SystemState {
         checkIntegerInRange("Group index", iGroup, 0, this._nGroups - 1);
         checkIntegerInRange("Particle index", iParticle, 0, this._nParticles - 1);
     }
+
+    /**
+     * Apply a reordering to the particles in every group according to a reordering parameter which defines
+     * a new order for each group.
+     * @param reordering {@link https://github.com/scijs/ndarray | NdArray } whose first dimension defines group index
+     * and whose second dimension defines a new order of particles for each group. For example, if this state has three
+     * particles per group, a reordering of `[2, 0, 1]` for any group would imply that its current third particle should
+     * be reordered to the first position, its current first particle to the second position, and its current second
+     * particle to the third position.
+     */
+    public reorder(reordering: ndarray.NdArray) {
+        const shape = reordering.shape;
+        if (shape.length !== 2 || shape[0] !== this._nGroups || shape[1] !== this._nParticles) {
+            throw new DustParameterError(
+                "Unexpected reordering shape. " +
+                    `Expected [${this._nGroups},${this._nParticles}] but got ${JSON.stringify(shape)}.`
+            );
+        }
+        // Check that each reordering contains expected values
+        for (let iGroup = 0; iGroup < this._nGroups; iGroup++) {
+            const reorder = particleStateToArray(reordering.pick(iGroup, null));
+            const sortedReorder = reorder.toSorted();
+            // Each reordering should contain every index value once so, when sorted, each value should equal its index
+            if (sortedReorder.some((value: number, index: number) => value !== index)) {
+                throw new DustParameterError(`Unexpected reorder values: ${JSON.stringify(reorder)}`);
+            }
```

Comment:
This is preventing sampling with replacement, no? The idea is not that we shuffle the values around, but that some will be duplicated and some omitted in the final list.  So reordering with a vector [0, 1, 1, 1, 5, 5] could be fine - we can't have values outside the range of `[0, n_particles-1]` but that's it

Reply:
Ah yeah, sorry, forgot about that requirement!